### PR TITLE
Windows CI: TP4 reliability hack

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

@mikedougherty @jfrazelle @icecrime @powersplay @friism @tonistiigi

This is a WAY better version of #19855. Still a horrible hack :smiling_imp:, but will make TP4 Windows to Windows CI far, far, far more reliable :heart:. I've also in this PR re-enabled the tests which were the most able to cause false failures, recently disabled in #19867. I've run the TestPsListContainers\* tests many many times locally on a TP4 machine and not seen failures after this hack is in place.

It also has the side effect of making docker more reliable for regular users using TP4 and a newer docker, once this is merged :smile:

### CODE REVIEW

**Comment by friism** [comment]
@jhowardmsft does this also remove the requirement to insert `sleep` stuff in the start of Windows Dockerfiles?

**Comment by jhowardmsft** [comment]
No, completely unrelated.

**Comment by calavera** [comment]
Can we do this when the driver is initialized to make it slightly less "hacky"? Having some retry logic seems reasonable.

**Comment by jhowardmsft** [comment]
@calavera Of course. Moved to driver init and validated locally.

**Comment by jhowardmsft** [comment]
10 runs of the tests which cause this the most often passed :smile:

```

E:\go\src\github.com\docker\docker>runtest w TestPsListContainers
Running test TestPsListContainers
DOCKER_TEST_HOST=tcp://127.0.0.1:2375
DOCKER_HOST=tcp://127.0.0.1:2375
=== RUN   Test
INFO: Testing against a local daemon
PASS: docker_cli_ps_test.go:19: DockerSuite.TestPsListContainersBase    15.697s
PASS: docker_cli_ps_test.go:247: DockerSuite.TestPsListContainersFilterAncestorImage    53.009s
PASS: docker_cli_ps_test.go:497: DockerSuite.TestPsListContainersFilterCreated  0.588s
PASS: docker_cli_ps_test.go:378: DockerSuite.TestPsListContainersFilterExited   21.819s
PASS: docker_cli_ps_test.go:209: DockerSuite.TestPsListContainersFilterID       7.652s
PASS: docker_cli_ps_test.go:342: DockerSuite.TestPsListContainersFilterLabel    11.510s
PASS: docker_cli_ps_test.go:224: DockerSuite.TestPsListContainersFilterName     7.596s
PASS: docker_cli_ps_test.go:170: DockerSuite.TestPsListContainersFilterStatus   8.601s
SKIP: docker_cli_ps_test.go:128: DockerSuite.TestPsListContainersSize (Test requires a Linux daemon)
SKIP: docker_cli_by_digest_test.go:340: DockerRegistrySuite.TestPsListContainersFilterAncestorImageByDigest
OK: 8 passed, 2 skipped
--- PASS: Test (140.67s)
PASS
ok      _/E_/go/src/github.com/docker/docker/integration-cli    140.877s

E:\go\src\github.com\docker\docker>runtest w TestPsListContainers
Running test TestPsListContainers
DOCKER_TEST_HOST=tcp://127.0.0.1:2375
DOCKER_HOST=tcp://127.0.0.1:2375
=== RUN   Test
INFO: Testing against a local daemon
PASS: docker_cli_ps_test.go:19: DockerSuite.TestPsListContainersBase    14.956s
PASS: docker_cli_ps_test.go:247: DockerSuite.TestPsListContainersFilterAncestorImage    42.447s
PASS: docker_cli_ps_test.go:497: DockerSuite.TestPsListContainersFilterCreated  0.602s
PASS: docker_cli_ps_test.go:378: DockerSuite.TestPsListContainersFilterExited   23.691s
PASS: docker_cli_ps_test.go:209: DockerSuite.TestPsListContainersFilterID       8.405s
PASS: docker_cli_ps_test.go:342: DockerSuite.TestPsListContainersFilterLabel    12.349s
PASS: docker_cli_ps_test.go:224: DockerSuite.TestPsListContainersFilterName     7.737s
PASS: docker_cli_ps_test.go:170: DockerSuite.TestPsListContainersFilterStatus   8.435s
SKIP: docker_cli_ps_test.go:128: DockerSuite.TestPsListContainersSize (Test requires a Linux daemon)
SKIP: docker_cli_by_digest_test.go:340: DockerRegistrySuite.TestPsListContainersFilterAncestorImageByDigest
OK: 8 passed, 2 skipped
--- PASS: Test (146.27s)
PASS
ok      _/E_/go/src/github.com/docker/docker/integration-cli    146.454s

E:\go\src\github.com\docker\docker>runtest w TestPsListContainers
Running test TestPsListContainers
DOCKER_TEST_HOST=tcp://127.0.0.1:2375
DOCKER_HOST=tcp://127.0.0.1:2375
=== RUN   Test
INFO: Testing against a local daemon
PASS: docker_cli_ps_test.go:19: DockerSuite.TestPsListContainersBase    17.023s
PASS: docker_cli_ps_test.go:247: DockerSuite.TestPsListContainersFilterAncestorImage    32.006s
PASS: docker_cli_ps_test.go:497: DockerSuite.TestPsListContainersFilterCreated  0.591s
PASS: docker_cli_ps_test.go:378: DockerSuite.TestPsListContainersFilterExited   22.750s
PASS: docker_cli_ps_test.go:209: DockerSuite.TestPsListContainersFilterID       6.897s
PASS: docker_cli_ps_test.go:342: DockerSuite.TestPsListContainersFilterLabel    11.527s
PASS: docker_cli_ps_test.go:224: DockerSuite.TestPsListContainersFilterName     7.913s
PASS: docker_cli_ps_test.go:170: DockerSuite.TestPsListContainersFilterStatus   8.588s
SKIP: docker_cli_ps_test.go:128: DockerSuite.TestPsListContainersSize (Test requires a Linux daemon)
SKIP: docker_cli_by_digest_test.go:340: DockerRegistrySuite.TestPsListContainersFilterAncestorImageByDigest
OK: 8 passed, 2 skipped
--- PASS: Test (122.09s)
PASS
ok      _/E_/go/src/github.com/docker/docker/integration-cli    122.211s

E:\go\src\github.com\docker\docker>runtest w TestPsListContainers
Running test TestPsListContainers
DOCKER_TEST_HOST=tcp://127.0.0.1:2375
DOCKER_HOST=tcp://127.0.0.1:2375
=== RUN   Test
INFO: Testing against a local daemon
PASS: docker_cli_ps_test.go:19: DockerSuite.TestPsListContainersBase    15.735s
PASS: docker_cli_ps_test.go:247: DockerSuite.TestPsListContainersFilterAncestorImage    43.611s
PASS: docker_cli_ps_test.go:497: DockerSuite.TestPsListContainersFilterCreated  0.667s
PASS: docker_cli_ps_test.go:378: DockerSuite.TestPsListContainersFilterExited   27.959s
PASS: docker_cli_ps_test.go:209: DockerSuite.TestPsListContainersFilterID       7.990s
PASS: docker_cli_ps_test.go:342: DockerSuite.TestPsListContainersFilterLabel    11.626s
PASS: docker_cli_ps_test.go:224: DockerSuite.TestPsListContainersFilterName     7.870s
PASS: docker_cli_ps_test.go:170: DockerSuite.TestPsListContainersFilterStatus   9.591s
SKIP: docker_cli_ps_test.go:128: DockerSuite.TestPsListContainersSize (Test requires a Linux daemon)
SKIP: docker_cli_by_digest_test.go:340: DockerRegistrySuite.TestPsListContainersFilterAncestorImageByDigest
OK: 8 passed, 2 skipped
--- PASS: Test (140.44s)
PASS
ok      _/E_/go/src/github.com/docker/docker/integration-cli    140.618s

E:\go\src\github.com\docker\docker>runtest w TestPsListContainers
Running test TestPsListContainers
DOCKER_TEST_HOST=tcp://127.0.0.1:2375
DOCKER_HOST=tcp://127.0.0.1:2375
=== RUN   Test
INFO: Testing against a local daemon
PASS: docker_cli_ps_test.go:19: DockerSuite.TestPsListContainersBase    15.536s
PASS: docker_cli_ps_test.go:247: DockerSuite.TestPsListContainersFilterAncestorImage    22.074s
PASS: docker_cli_ps_test.go:497: DockerSuite.TestPsListContainersFilterCreated  0.617s
PASS: docker_cli_ps_test.go:378: DockerSuite.TestPsListContainersFilterExited   23.262s
PASS: docker_cli_ps_test.go:209: DockerSuite.TestPsListContainersFilterID       7.933s
PASS: docker_cli_ps_test.go:342: DockerSuite.TestPsListContainersFilterLabel    11.540s
PASS: docker_cli_ps_test.go:224: DockerSuite.TestPsListContainersFilterName     7.809s
PASS: docker_cli_ps_test.go:170: DockerSuite.TestPsListContainersFilterStatus   8.776s
SKIP: docker_cli_ps_test.go:128: DockerSuite.TestPsListContainersSize (Test requires a Linux daemon)
SKIP: docker_cli_by_digest_test.go:340: DockerRegistrySuite.TestPsListContainersFilterAncestorImageByDigest
OK: 8 passed, 2 skipped
--- PASS: Test (113.27s)
PASS
ok      _/E_/go/src/github.com/docker/docker/integration-cli    113.513s

E:\go\src\github.com\docker\docker>runtest w TestPsListContainers
Running test TestPsListContainers
DOCKER_TEST_HOST=tcp://127.0.0.1:2375
DOCKER_HOST=tcp://127.0.0.1:2375
=== RUN   Test
INFO: Testing against a local daemon
PASS: docker_cli_ps_test.go:19: DockerSuite.TestPsListContainersBase    15.811s
PASS: docker_cli_ps_test.go:247: DockerSuite.TestPsListContainersFilterAncestorImage    43.078s
PASS: docker_cli_ps_test.go:497: DockerSuite.TestPsListContainersFilterCreated  0.605s
PASS: docker_cli_ps_test.go:378: DockerSuite.TestPsListContainersFilterExited   24.203s
PASS: docker_cli_ps_test.go:209: DockerSuite.TestPsListContainersFilterID       8.334s
PASS: docker_cli_ps_test.go:342: DockerSuite.TestPsListContainersFilterLabel    12.502s
PASS: docker_cli_ps_test.go:224: DockerSuite.TestPsListContainersFilterName     7.389s
PASS: docker_cli_ps_test.go:170: DockerSuite.TestPsListContainersFilterStatus   8.657s
SKIP: docker_cli_ps_test.go:128: DockerSuite.TestPsListContainersSize (Test requires a Linux daemon)
SKIP: docker_cli_by_digest_test.go:340: DockerRegistrySuite.TestPsListContainersFilterAncestorImageByDigest
OK: 8 passed, 2 skipped
--- PASS: Test (135.84s)
PASS
ok      _/E_/go/src/github.com/docker/docker/integration-cli    136.039s

E:\go\src\github.com\docker\docker>runtest w TestPsListContainers
Running test TestPsListContainers
DOCKER_TEST_HOST=tcp://127.0.0.1:2375
DOCKER_HOST=tcp://127.0.0.1:2375
=== RUN   Test
INFO: Testing against a local daemon
PASS: docker_cli_ps_test.go:19: DockerSuite.TestPsListContainersBase    15.732s
PASS: docker_cli_ps_test.go:247: DockerSuite.TestPsListContainersFilterAncestorImage    33.190s
PASS: docker_cli_ps_test.go:497: DockerSuite.TestPsListContainersFilterCreated  0.604s
PASS: docker_cli_ps_test.go:378: DockerSuite.TestPsListContainersFilterExited   24.632s
PASS: docker_cli_ps_test.go:209: DockerSuite.TestPsListContainersFilterID       8.184s
PASS: docker_cli_ps_test.go:342: DockerSuite.TestPsListContainersFilterLabel    13.625s
PASS: docker_cli_ps_test.go:224: DockerSuite.TestPsListContainersFilterName     8.142s
PASS: docker_cli_ps_test.go:170: DockerSuite.TestPsListContainersFilterStatus   8.588s
SKIP: docker_cli_ps_test.go:128: DockerSuite.TestPsListContainersSize (Test requires a Linux daemon)
SKIP: docker_cli_by_digest_test.go:340: DockerRegistrySuite.TestPsListContainersFilterAncestorImageByDigest
OK: 8 passed, 2 skipped
--- PASS: Test (128.10s)
PASS
ok      _/E_/go/src/github.com/docker/docker/integration-cli    128.301s

E:\go\src\github.com\docker\docker>runtest w TestPsListContainers
Running test TestPsListContainers
DOCKER_TEST_HOST=tcp://127.0.0.1:2375
DOCKER_HOST=tcp://127.0.0.1:2375
=== RUN   Test
INFO: Testing against a local daemon
PASS: docker_cli_ps_test.go:19: DockerSuite.TestPsListContainersBase    16.476s
PASS: docker_cli_ps_test.go:247: DockerSuite.TestPsListContainersFilterAncestorImage    53.740s
PASS: docker_cli_ps_test.go:497: DockerSuite.TestPsListContainersFilterCreated  0.621s
PASS: docker_cli_ps_test.go:378: DockerSuite.TestPsListContainersFilterExited   24.519s
PASS: docker_cli_ps_test.go:209: DockerSuite.TestPsListContainersFilterID       8.212s
PASS: docker_cli_ps_test.go:342: DockerSuite.TestPsListContainersFilterLabel    12.629s
PASS: docker_cli_ps_test.go:224: DockerSuite.TestPsListContainersFilterName     7.602s
PASS: docker_cli_ps_test.go:170: DockerSuite.TestPsListContainersFilterStatus   9.254s
SKIP: docker_cli_ps_test.go:128: DockerSuite.TestPsListContainersSize (Test requires a Linux daemon)
SKIP: docker_cli_by_digest_test.go:340: DockerRegistrySuite.TestPsListContainersFilterAncestorImageByDigest
OK: 8 passed, 2 skipped
--- PASS: Test (148.20s)
PASS
ok      _/E_/go/src/github.com/docker/docker/integration-cli    148.408s

E:\go\src\github.com\docker\docker>runtest w TestPsListContainers
Running test TestPsListContainers
DOCKER_TEST_HOST=tcp://127.0.0.1:2375
DOCKER_HOST=tcp://127.0.0.1:2375
=== RUN   Test
INFO: Testing against a local daemon
PASS: docker_cli_ps_test.go:19: DockerSuite.TestPsListContainersBase    17.422s
PASS: docker_cli_ps_test.go:247: DockerSuite.TestPsListContainersFilterAncestorImage    33.367s
PASS: docker_cli_ps_test.go:497: DockerSuite.TestPsListContainersFilterCreated  0.611s
PASS: docker_cli_ps_test.go:378: DockerSuite.TestPsListContainersFilterExited   23.962s
PASS: docker_cli_ps_test.go:209: DockerSuite.TestPsListContainersFilterID       8.698s
PASS: docker_cli_ps_test.go:342: DockerSuite.TestPsListContainersFilterLabel    14.097s
PASS: docker_cli_ps_test.go:224: DockerSuite.TestPsListContainersFilterName     10.033s
PASS: docker_cli_ps_test.go:170: DockerSuite.TestPsListContainersFilterStatus   8.816s
SKIP: docker_cli_ps_test.go:128: DockerSuite.TestPsListContainersSize (Test requires a Linux daemon)
SKIP: docker_cli_by_digest_test.go:340: DockerRegistrySuite.TestPsListContainersFilterAncestorImageByDigest
OK: 8 passed, 2 skipped
--- PASS: Test (132.01s)
PASS
ok      _/E_/go/src/github.com/docker/docker/integration-cli    132.291s

E:\go\src\github.com\docker\docker>runtest w TestPsListContainers
Running test TestPsListContainers
DOCKER_TEST_HOST=tcp://127.0.0.1:2375
DOCKER_HOST=tcp://127.0.0.1:2375
=== RUN   Test
INFO: Testing against a local daemon
PASS: docker_cli_ps_test.go:19: DockerSuite.TestPsListContainersBase    17.061s
PASS: docker_cli_ps_test.go:247: DockerSuite.TestPsListContainersFilterAncestorImage    42.783s
PASS: docker_cli_ps_test.go:497: DockerSuite.TestPsListContainersFilterCreated  0.599s
PASS: docker_cli_ps_test.go:378: DockerSuite.TestPsListContainersFilterExited   23.404s
PASS: docker_cli_ps_test.go:209: DockerSuite.TestPsListContainersFilterID       7.267s
PASS: docker_cli_ps_test.go:342: DockerSuite.TestPsListContainersFilterLabel    11.370s
PASS: docker_cli_ps_test.go:224: DockerSuite.TestPsListContainersFilterName     9.443s
PASS: docker_cli_ps_test.go:170: DockerSuite.TestPsListContainersFilterStatus   8.521s
SKIP: docker_cli_ps_test.go:128: DockerSuite.TestPsListContainersSize (Test requires a Linux daemon)
SKIP: docker_cli_by_digest_test.go:340: DockerRegistrySuite.TestPsListContainersFilterAncestorImageByDigest
OK: 8 passed, 2 skipped
--- PASS: Test (135.02s)
PASS
ok      _/E_/go/src/github.com/docker/docker/integration-cli    135.246s

E:\go\src\github.com\docker\docker>
```

**Comment by mikedougherty** [comment]
i don't know much about the code but i like the results (both from John's testing and all the automated builds) so LGTM :)

**Comment by icecrime** [comment]
LGTM :-)

File: daemon/execdriver/windows/run.go [hunk 1]
```diff
@@ -223,10 +224,31 @@ func (d *Driver) Run(c *execdriver.Command, pipes *execdriver.Pipes, hooks execd
 
 	configuration := string(configurationb)
 
-	err = hcsshim.CreateComputeSystem(c.ID, configuration)
-	if err != nil {
-		logrus.Debugln("Failed to create temporary container ", err)
-		return execdriver.ExitStatus{ExitCode: -1}, err
+	// The following a workaround for Windows TP4 which has a networking
+	// bug which fairly frequently returns an error. Back off and retry.
```
Comment (tiborvass) [inline daemon/execdriver/windows/run.go]: would be great to add a `TODO(windows)` to make sure this hack gets removed in time :)

**Comment by tiborvass** [comment]
Apart from small nit, LGTM

File: daemon/execdriver/windows/run.go (hunk 1 above)
Comment (jhowardmsft) [inline daemon/execdriver/windows/run.go]: @tiborvass Added.

**Comment by jhowardmsft** [comment]
@calavera Can this move to code review based on the above LGTMs? :smile: Thanks!

**Comment by jhowardmsft** [comment]
Yay, thank you! That will make a huge difference to windowsTP4 context reliability! Now to update the boxes with a control daemon including this fix....

**Pull request merged**